nemo/utils/model_utils.py

```diff
@@ -0,0 +1,237 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import os
+from pathlib import Path
+
+from omegaconf import DictConfig, ListConfig, OmegaConf
+
+from nemo import logging
+
+
+def resolve_dataset_name_from_cfg(cfg: DictConfig) -> str:
+    """
+    Parses items of the provided sub-config to find the first potential key that
+    resolves to an existing file or directory.
+
+    NOTE:
+    It <can> potentially mismatch if there exist more than 2 valid paths, and the
+    first path does *not* resolve the the path of the data file (but does resolve to
+    some other valid path).
+
+    To avoid this side-effect, place the data path as the first item on the config file.
+
+    Args:
+        cfg: DictConfig (Sub-config) that should be parsed.
+
+    Returns:
+        A str representing the `key` of the config which hosts the filepath(s),
+        or None in case path could not be resolved.
+    """
+    if hasattr(cfg, 'ds_name') and cfg.ds_name is not None:
+        return 'ds_name'
+
+    for key, value in cfg.items():
+        if type(value) in [list, tuple, ListConfig]:
+            # Count the number of valid paths in the list
+            values_are_paths = 0
+            for val_i in value:
+                val_i = str(val_i)
+
+                if os.path.exists(val_i) or os.path.isdir(val_i):
+                    values_are_paths += 1
+                else:
+                    # reset counter and break inner loop
+                    break
+
+            if values_are_paths == len(value):
+                return key
+
+        else:
+            if os.path.exists(str(value)) or os.path.isdir(str(value)):
+                return key
+
+    return None
+
+
+def parse_dataset_as_name(name: str) -> str:
+    """
+    Constructs a valid prefix-name from a provided file path.
+
+    Args:
+        name: str path to some valid data/manifest file or a python object that
+            will be used as a name for the data loader (via str() cast).
+
+    Returns:
+        str prefix used to identify uniquely this data/manifest file.
+    """
+    if os.path.exists(str(name)) or os.path.isdir(str(name)):
+        name = Path(name).stem
+    else:
+        name = str(name)
+
+    # cleanup name
+    name = name.replace('-', '_')
+
+    if 'manifest' in name:
+        name = name.replace('manifest', '')
+
+    if 'dataset' in name:
+        name = name.replace('dataset', '')
+
+    if '_' != name[-1]:
+        name = name + '_'
+
+    return name
+
+
+def resolve_validation_dataloaders(model: 'ModelPT'):
+    """
+    Helper method that operates on the ModelPT class to automatically support
+    multiple dataloaders for the validation set.
+
+    It does so by first resolving the path to one/more data files via `resolve_filepath_from_cfg()`.
```

Comment:
should `resolve_filepath_from_cfg` be `resolve_dataset_name_from_cfg`?

Reply:
Done